synapse/util/caches/dictionary_cache.py

```diff
@@ -96,28 +145,116 @@ def get(
+            if entry is _Sentinel.sentinel:
+                missing.add(dict_key)
+                continue
+
+            assert isinstance(entry, _PerKeyValue)
+
+            if entry.value is _Sentinel.sentinel:
+                known_absent.add(dict_key)
+            else:
+                values[dict_key] = entry.value
+
+        # If we found everything we can return immediately.
+        if not missing:
+            return DictionaryEntry(False, known_absent, values)
+
+        # If we are missing any keys check if we happen to have the full dict in
+        # the cache.
+        #
+        # We don't update the last access time for this cache fetch, as we
+        # aren't explicitly interested in the full dict and so we don't want
+        # requests for explicit dict keys to keep the full dict in the cache.
+        entry = self.cache.get(
+            (key, _FullCacheKey.KEY),
+            _Sentinel.sentinel,
+            update_last_access=False,
+        )
+        if entry is _Sentinel.sentinel:
+            # Not in the cache, return the subset of keys we found.
+            return DictionaryEntry(False, known_absent, values)
+
+        # We have the full dict!
+        assert isinstance(entry, dict)
+
+        values = {}
+        for dict_key in dict_keys:
+            # We explicitly add each dict key to the cache, so that cache hit
+            # rates for each key can be tracked separately.
+            value = entry.get(dict_key, _Sentinel.sentinel)  # type: ignore[arg-type]
+            self.cache[(key, dict_key)] = _PerKeyValue(value)
+
+            if value is not _Sentinel.sentinel:
+                values[dict_key] = value
+
+        return DictionaryEntry(True, set(), values)
+
+    def _get_full_dict(
+        self,
+        key: KT,
+    ) -> DictionaryEntry:
+        """Fetch the full dict for the given key."""
+
+        # First we check if we have cached the full dict.
+        entry = self.cache.get((key, _FullCacheKey.KEY), _Sentinel.sentinel)
         if entry is not _Sentinel.sentinel:
-            if dict_keys is None:
-                return DictionaryEntry(
-                    entry.full, entry.known_absent, dict(entry.value)
-                )
+            assert isinstance(entry, dict)
+            return DictionaryEntry(True, set(), entry)
+
+        # If not, check if we have cached any dict keys at all for this cache
+        # key.
+        all_entries = self.cache.get_multi(
+            (key,),
+            _Sentinel.sentinel,
+        )
+        if all_entries is _Sentinel.sentinel:
+            return DictionaryEntry(False, set(), {})
+
+        # If there are entries we need to unwrap the returned cache nodes
+        # and `_PerKeyValue` into the `DictionaryEntry`.
+        values = {}
+        known_absent = set()
+        for cache_key, dict_value in all_entries:
+            # The key used for the `TreeCache` is `(key, dict_key)`
```

Comment:
at this point we don't really care that `self.cache` is backed by a TreeCache.

```suggestion
            # The keys in `self.cache` are `(cache_key, dict_key)`
```